### PR TITLE
Tab completion NFA DSL

### PR DESCRIPTION
**Pull request by georgebrock**
#296 introduced a non-deterministic finite automaton as the basis of the tab completion system. This PR is the next step in my grand plan for tab completion: it introduces a DSL for loading the NFA's state graph from a configuration file.

The DSL format is described in the `gitsh_completions.5` manual page, which is added in this PR.

The configuration file included in this PR (`etc/completions`) results in a graph very similar to the one that was hard coded prior to the introduction of the DSL. Subsequent PRs will add features to the DSL so it can be used for more complex graphs that produce better tab completion behaviour.

Note that I'm not using DSL in the sense it's usually used with Ruby: this language has a lexer and a parser, it's not just a pretty looking Ruby API.

### CODE REVIEW

File: lib/gitsh/tab_completion/dsl.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+module Gitsh
+  module TabCompletion
+    module DSL
+      def self.load(path, start_state, env)
+        source = File.read(path)
+        tokens = Lexer.lex(source)
+        factory = Parser.parse(tokens, gitsh_env: env)
+        factory.build(start_state)
+      rescue Errno::ENOENT
+      end
```
Comment (mike-burns) [inline lib/gitsh/tab_completion/dsl.rb]: The silence here makes me suspect that we want a conditional elsewhere instead. I feel like `DSL.load` should complain about a missing file, but that `DSL.load` should only be called if we want such a complaint ... if that makes sense.

File: lib/gitsh/tab_completion/dsl/concatenation_factory.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      class ConcatenationFactory
+        attr_reader :parts
+
+        def initialize(parts)
+          @parts = parts
+        end
+
+        def build(start_state, options = {})
+          with_optional_end_state(options) do
+            parts.inject(start_state) do |state, part|
```
Comment (mike-burns) [inline lib/gitsh/tab_completion/dsl/concatenation_factory.rb]: > But that's not what I'm here to tell you about.
> I'm here to talk about the draft.
> They got a buildin' down in New York City called Whitehall Street, where you 
> Walk in, you get injected, inspected, detected, infected, neglected and 
> Selected!

File: man/man1/gitsh.1.in [hunk 3]
```diff
@@ -406,6 +406,13 @@ The
 .Pa .gitshrc
 file will not be loaded for non-interactive sessions, e.g. when gitsh is
 invoked with the path to a script file.
+.It Pa @prefix@/etc/gitsh/completions
```
Comment (mike-burns) [inline man/man1/gitsh.1.in]: Maybe `@pkgsysconfdir@` instead of `@prefix@` is more accurate.

File: man/man5/gitsh_completions.5.in [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+files define the tab completions options that are available in interactive
+.Xr gitsh 1
+sessions.
+.Pp
+The content of a
+.Nm gitsh_completions
+file consists of rules defining the arguments that are expected
+by various commands.
+.Xr gitsh 1
+uses the rules to match the input before the word being completed,
```
Comment (mike-burns) [inline man/man5/gitsh_completions.5.in]: Probably no comma.

File: man/man5/gitsh_completions.5.in [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+and to generate completions that are relevant in the context.
+.Pp
+Each rule is separated from other rules by a blank line.
+.Pp
+The rule specifies the name of the command,
```
Comment (mike-burns) [inline man/man5/gitsh_completions.5.in]: No comma.

File: spec/support/tab_completion.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+module TabCompletionHelpers
+  def stub_text_matcher(text)
+    klass = Gitsh::TabCompletion::Matchers::TextMatcher
+    matcher = instance_double(klass)
+    allow(klass).to receive(:new).with(text).and_return(matcher)
```
Comment (mike-burns) [inline spec/support/tab_completion.rb]: `#with` makes this into a mock instead of a stub, right?

File: spec/units/tab_completion/dsl_spec.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        env = Gitsh::Environment.new
+
+        expect { described_class.load(path, start_state, env) }.
+          not_to raise_exception
+      end
+    end
+  end
+
+  def write_temp_completions_file(completions)
+    temp_file('gitsh_completions', completions).path
```
Comment (mike-burns) [inline spec/units/tab_completion/dsl_spec.rb]: How's this work -- do these files get cleaned up?

**Comment by georgebrock** [comment]
> I don't love `GITSH_CONFIG_DIRECTORY`, but I also don't want to make you pass a string through a series of method calls.

It might not be too bad: the file it's current defined in would pass it to `Gitsh::CLI`, to `Gitsh::InputStrategies::Interactive`, to `Gitsh::TabCompletion::Facade`, to `Gitsh::TabCompletion::AutomatonFactory` where it's used.

File: lib/gitsh/tab_completion/dsl.rb (hunk 1 above)
Comment (georgebrock) [inline lib/gitsh/tab_completion/dsl.rb]: I could move the rescue up a level to the `Gitsh::TabCompletion::AutomatonFactory` object?

I don't think I want a missing configuration file to cause an error that the user sees, but maybe it should—without the file there will be no tab completion.

File: man/man1/gitsh.1.in (hunk 3 above)
Comment (georgebrock) [inline man/man1/gitsh.1.in]: Good call.

File: spec/units/tab_completion/dsl_spec.rb (hunk 7 above)
Comment (georgebrock) [inline spec/units/tab_completion/dsl_spec.rb]: Yes. It uses the Ruby standard library's `Tempfile` class behind the scenes, so it should be cleaned up when the object is garbage collected.

File: spec/support/tab_completion.rb (hunk 6 above)
Comment (georgebrock) [inline spec/support/tab_completion.rb]: I never get those distinctions right, but you're probably right.

**Comment by georgebrock** [comment]
Thinking about this configuration thing some more: we already pass the `Gitsh::Environment` instance around all of these places, and it's already able to hold variables. If we put this path in a variable, that would allow people to override it in their `.gitshrc` file.

File: lib/gitsh/tab_completion/dsl/concatenation_factory.rb (hunk 2 above)
Comment (georgebrock) [inline lib/gitsh/tab_completion/dsl/concatenation_factory.rb]: Hound config updated.

File: spec/units/tab_completion/dsl/rule_set_factory_spec.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
+require 'spec_helper'
+require 'gitsh/tab_completion/dsl/rule_set_factory'
+require 'gitsh/tab_completion/dsl/rule_factory'
+
+describe Gitsh::TabCompletion::DSL::RuleSetFactory do
+  describe '#build' do
+    it 'delegates to the various rule factories' do
+      start_state = double(:start_state)
+      rule_factory_1 = stub_rule_factory
```
Comment (georgebrock) [inline spec/units/tab_completion/dsl/rule_set_factory_spec.rb]: Hound config updated.

File: spec/units/tab_completion/dsl/parser_spec.rb [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    it 'parses single variables' do
+      result = parse_single_rule(tokens([:VAR, 'revision'], [:EOS]))
+
+      expect(result).to be_a_variable_transition
+      expect(result.matcher).to be_a_revision_matcher
+    end
+
+    it 'parses rules with multiple words and variables' do
+      result = parse_single_rule(tokens(
+        [:WORD, 'stash'], [:WORD, 'pop'], [:VAR, 'revision'], [:EOS],
```
Comment (georgebrock) [inline spec/units/tab_completion/dsl/parser_spec.rb]: Hound config updated for the trailing commas.

Comment (georgebrock) [inline spec/units/tab_completion/dsl/parser_spec.rb]: The indentation comment seems to be an example of this issue, so I guess we'll just live with that one: https://github.com/bbatsov/rubocop/issues/4308

File: spec/units/tab_completion/dsl_spec.rb [hunk 10]
```diff
@@ -0,0 +1,37 @@
+require 'spec_helper'
+require 'gitsh/tab_completion/dsl'
+
+describe Gitsh::TabCompletion::DSL do
+  describe '.load' do
+    it 'builds a graph of automaton states by reading the given file' do
+      path = write_temp_completions_file([
+        'stash (apply|drop|pop|show)',
+      ].join("\n"))
```
Comment (georgebrock) [inline spec/units/tab_completion/dsl_spec.rb]: Err … this isn't quite what I intended. Rubocop seems to struggle with nested multiline things.

**Comment by georgebrock** [comment]
@mike-burns I think 3a76b3fbc6e806244d799fb43dac49dd7f087168 addresses your concerns about the global constant.

**Comment by mike-burns** [comment]
Sure does!

LGTM.

**Pull request merged**